net: arp: Add support for receiving gratuitous ARP request

jukkar opened the pull request:
If we receive a valid gratuitous ARP request, then update ARP
cache accordingly. This feature is optional and by default
it is enabled, but can be turned off if needed.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: arp: Add support for receiving gratuitous ARP request

If we receive a valid gratuitous ARP request, then update ARP
cache accordingly. This feature is optional and by default
it is enabled, but can be turned off if needed.

This is similar to a feature available in Linux, and is enabled
by default in many distros.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

pfalcon commented:
Weird, this didn't have automatic reviewers assignment.

pfalcon commented:
What's the reason for implementing this?

"Gratuitous" things are known common attack vector, so I'm not sure about on by default.

jukkar commented:
> What's the reason for implementing this?

This is related to issue #10188 

It is certainly possible to use gratuitous ARP as an attack vector. This seems to be more issue in 802.11 (wifi) networks that we currently do not support. Linux enables this by default, in my Fedora installation the `cat /proc/sys/net/ipv4/conf/all/drop_gratuitous_arp` returns 0 which means that gratuitous ARPs are accepted.

See also relevant linux source code at https://github.com/torvalds/linux/blob/64c5e530ac2cab7e710c3bd08ef8594a016e4f9a/net/ipv4/arp.c#L772

tbursztyka commented on subsys/net/l2/ethernet/arp.c:
```diff
@@ -445,9 +445,31 @@ struct net_pkt *net_arp_prepare(struct net_pkt *pkt,
-static inline void arp_update(struct net_if *iface,
-			      struct in_addr *src,
-			      struct net_eth_addr *hwaddr)
+static void gratuitous_arp(struct net_if *iface,
```
can you rename it to arp_gratuitous ?

jukkar replied:
That sounds very awkward actually as typically this support is called "gratuitous ARP", so naming the function same way is quite practical.

jukkar replied:
We can certainly change the name, I have no issues with that. It is just I do not fully agree this naming convention.

pfalcon replied:
That's one of those "d'oh" moments. @tbursztyka created half a dozen of of "cleanup" tickets requesting following such a naming convention in various modules, and we've been working on them for a few months already. And now turns out we don't even have agreement on that naming convention.

Anyway, trying to find a way forward: you could name it arp_gratuitous_arp(), arp_gratuitous_req(), arp_gratuitous_msg() but IMHO those aren't really improvements over just arp_gratuitous().

jukkar replied:
> trying to find a way forward: you could name it arp_gratuitous_arp(), arp_gratuitous_req(), arp_gratuitous_msg() but IMHO those aren't really improvements over just arp_gratuitous().

No need for this madness, the function name can be set to `arp_gratuitous()` if needed, this is minor issue. I just do not see the point in this naming convention.

pfalcon commented:
> in my Fedora installation the cat /proc/sys/net/ipv4/conf/all/drop_gratuitous_arp returns 0

Same here on Ubuntu 18.04.

Thanks for explanations, ideally, at least some of them would go into the commit message (along the lines "this is similar to feature available in Linux, and enabled by default in many distros").

Looks good for me one the function name is made consistent with the existing funcs in the file.

jukkar committed:
tests: net: arp: Add gratuitous ARP tests

Make sure that we change the HW address in ARP cache if we
receive gratuitous ARP.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

pfalcon reviewed (approved):
Thanks!
